### DISCUSSION TITLE
Next.js API functions run multiple times with different results

### DISCUSSION THREAD

**Discussion by ilyador** [vercel/next.js#43580] [Help] [answered]
I am using a Next.js API call to retrieve some data from an external API (I know this is better done with getStaticProps but I need to make a few calls before passing the data to the client.)
I don't understand why doe the backend function runs multiple times, and on the first time it doesn't have the POST data.

cleint
```javascript
export default function Customer () {
  const router = useRouter()
  const { id: submissionId } = router.query

  const { data: customer, error } = useSWR(['/api/jotform/get-client/', submissionId], fetcher)
...
```

server 
```javascript
export default async function handler (req, res) {
  try {
    const customers = await axios.get('https://api.jotform.com/form/' + jotformAPI.form_id + '/submissions?apiKey=' + jotformAPI.key)

    const { submissionId } = req.body
    const customer = customers.data.content.find(client => client.id === submissionId)
    if (!customer) throw new Error('contact does not exist')
    //on the first call the error is triggered, only on the second time I get the data.
...
```

**Comment by icyJoseph** [comment, marked as answer]
## Edit + Solution

The issue lies within router.query behaviour:

> query: Object - The query string parsed to an object, including [dynamic route](https://nextjs.org/docs/routing/dynamic-routes) parameters. It will be an empty object during prerendering if the page doesn't use [Server-side Rendering](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props). Defaults to {}

So if the page doesn't use GetServerSideProps for example, then it'll first render with query empty, and then receive an update where the query contains the values.

https://nextjs.org/docs/api-reference/next/router

![Screenshot 2022-12-01 at 17 20 03](https://user-images.githubusercontent.com/21013447/205104996-7fd24768-4334-4671-9217-84f0b6001b27.png)

In this case, because SWR needs a value from the query, we can use conditional fetching, https://swr.vercel.app/docs/conditional-fetching#conditional:

```jsx
import Head from "next/head";
import { useRouter } from "next/router";
import useSWR from "swr";

const fetcher = (...args: unknown[]) => {
  console.log(args);
  return fetch("/api/hello");
};

export default function Home() {
  const { query } = useRouter();

  const id = query.id;

  const { data } = useSWR(id ? ["customer", id] : null, fetcher);

  // ❗️❌ For SEO this not very good! ❗️❌
  if (!data) return <div>Loading...</div>;

  return (
    <div>
      <Head>
        <title>Home title</title>
      </Head>

      <h1>Home</h1>
    </div>
  );
}

```

And now the fetcher is called only once.

## Why is query empty on first frame?

> Pages that are statically optimized by [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization) will be hydrated without their route parameters provided, i.e query will be an empty object ({}).
> 
> After hydration, Next.js will trigger an update to your application to provide the route parameters in the query object.

Pages that'll be opted-in to pre-rendering cannot possibly have knowledge of the query. Because pre-rendering happens on a server, without a request incoming... so there's no query to base ourselves of.

**Reply by ilyador** [reply]
Both times POST

**Reply by icyJoseph** [reply]
Ah sorry I missed the router query and swr... During first render query is empty. So you gotta tell swr to conditionally use the fetcher. I'll be home in a few minutes, then update my answer with code.

I thought this was a preflight issue, but all you need to do is to tweak how you use swr. 

https://swr.vercel.app/docs/conditional-fetching#conditional

**Reply by icyJoseph** [reply]
@ilyador Alright now I've updated the main answer, please do confirm if it fixes your issue.

**Reply by ilyador** [reply]
@icyJoseph yes it does, thank you.
BTW why does the part you marked not good for SEO?

Is `null` required in the conditional call or can it be shorthanded to
```javascript
useSWR(id && ["customer", id], fetcher);
```

**Reply by icyJoseph** [reply]
By itself is not necessarily bad for SEO.

The thing is that the SSR HTML string will contain the loading branch. It'll only ever update if JS is available.

So you better hope the SEO tool runs JS, which really depends on each bot/crawler. Most just curl your site and collect the SSR HTML.

There's some gotchas here that I'll skip over.
